horovod/run/run.py

```diff
@@ -714,28 +789,140 @@ def run():
+        return None
+
+
+def _launch_job(args, remote_host_names, settings, common_intfs, command):
     env = os.environ.copy()
     config_parser.set_env_from_args(env, args)
+    driver_ip = _get_driver_ip(common_intfs)
 
     if args.use_gloo:
         if not gloo_built(verbose=(settings.verbose >= 2)):
             raise ValueError('Gloo support has not been built.  If this is not expected, ensure CMake is installed '
                              'and reinstall Horovod with HOROVOD_WITH_GLOO=1 to debug the build error.')
-        gloo_run(settings, remote_host_names, common_intfs, env)
+        gloo_run(settings, remote_host_names, common_intfs, env, driver_ip, command)
     elif args.use_mpi:
         if not mpi_built(verbose=(settings.verbose >= 2)):
             raise ValueError('MPI support has not been built.  If this is not expected, ensure MPI is installed '
                              'and reinstall Horovod with HOROVOD_WITH_MPI=1 to debug the build error.')
-        mpi_run(settings, common_intfs, env)
+        mpi_run(settings, common_intfs, env, command)
     else:
         if mpi_built(verbose=(settings.verbose >= 2)):
-            mpi_run(settings, common_intfs, env)
+            mpi_run(settings, common_intfs, env, command)
         elif gloo_built(verbose=(settings.verbose >= 2)):
-            gloo_run(settings, remote_host_names, common_intfs, env)
+            gloo_run(settings, remote_host_names, common_intfs, env, driver_ip, command)
         else:
             raise ValueError('Neither MPI nor Gloo support has been built. Try reinstalling Horovod ensuring that '
                              'either MPI is installed (MPI) or CMake is installed (Gloo).')
 
 
+def run_commandline():
+    args = parse_args()
+    args.run_func = None
+    _run(args)
+
+
+def run(
+        func,
+        args=(),
+        kwargs={},
+        np=1,
+        hosts=None,
+        hostfile=None,
+        start_timeout=None,
+        ssh_port=None,
+        disable_cache=None,
+        output_filename=None,
+        verbose=None,
+        use_gloo=None,
+        use_mpi=None):
+    """
+    Launch a Horovod job to run the specified process function and get the return value.
+
+    :param func: The function to be run in Horovod job processes. The function return value will
+                 be collected as the corresponding Horovod process return value.
+    :param args: Arguments to pass to `func`.
+    :param kwargs: Keyword arguments to pass to `func`.
+    :param np: Number of Horovod processes.
+    :param hosts: List of host names and the number of available slots
+                  for running processes on each, of the form: <hostname>:<slots>
+                  (e.g.: host1:2,host2:4,host3:1 indicating 2 processes can run on host1,
+                  4 on host2, and 1 on host3). If not specified, defaults to using localhost:<np>
+    :param hostfile: Path to a host file containing the list of host names and the number of
```

Comment:
Do we allow both `hosts` and `hostfile`? If both are provided, what is the behavior?

Reply:
Do not allowed providing both. See `parse_args` method there's exclude rule checking.
But here I can add additional checking logic